core/src/main/java/io/grpc/internal/SingleProducerSerializingExecutor.java

```diff
@@ -0,0 +1,224 @@
+ * #execute} may not be called concurrently by multiple threads.
+ */
+@NotThreadSafe
+public final class SingleProducerSerializingExecutor implements Executor {
+  private static final Logger log =
+      Logger.getLogger(SerializingExecutor.class.getName());
+  private static final
+      AtomicReferenceFieldUpdater<SingleProducerSerializingExecutor, LinkedRunnable> tailUpdater =
+      AtomicReferenceFieldUpdater.newUpdater(
+          SingleProducerSerializingExecutor.class, LinkedRunnable.class, "tail");
+
+  /** Underlying executor that all submitted Runnable objects are run on. */
+  private final Executor executor;
+  /**
+   * If {@code null}, TaskRunner is not scheduled or running. This is the actual tail of the queue,
+   * independent of whether {@link LinkedRunnable#next} is non-{@code null}.
+   *
+   * <p>Although it might seem like we could use a single-producer, single-consumer queue, it's a
+   * bit more complex than that because we also have to manage starting and restarting the consumer.
+   * So we still require the producer and consumer to write to a single shared location.
+   */
+  private volatile LinkedRunnable tail;
+  /** The object that actually runs the Runnables submitted, reused. */
+  private final TaskRunner taskRunner = new TaskRunner();
+
+  /**
+   * Creates a SerializingExecutor, running tasks using {@code executor}.
+   *
+   * @param executor Executor in which tasks should be run. Must not be null.
+   */
+  public SingleProducerSerializingExecutor(Executor executor) {
+    this.executor = Preconditions.checkNotNull(executor, "executor");
+  }
+
+  private boolean tailCompareAndSet(LinkedRunnable expect, LinkedRunnable update) {
+    return tailUpdater.compareAndSet(this, expect, update);
+  }
+
+  /**
+   * Runs the given runnable strictly after all Runnables that were submitted before it, and using
+   * the {@code executor} passed to the constructor.
+   */
+  @Override
+  public void execute(Runnable r) {
+    Preconditions.checkNotNull(r, "runnable");
+    while (true) {
+      LinkedRunnable cachedTail = tail;
+      if (cachedTail == null) { // TaskRunnable not running.
+        cachedTail = new LinkedRunnable(r);
+        if (!tailCompareAndSet(null, cachedTail)) {
+          // Impossible, since TaskRunnable isn't running and we assume single producer.
+          throw new AssertionError();
+        }
+        taskRunner.schedule(cachedTail);
+        return;
+      } else {
+        // Must set 'next' before changing 'tail', so that the volatile 'write' to tail publishes
+        // the write to 'next'. It also makes the LinkedRunnable available to TaskRunner immediately
+        // when it discovers tail has changed.
+        cachedTail.next = new LinkedRunnable(r);
+        if (!tailCompareAndSet(cachedTail, cachedTail.next)) {
+          // 'next' is guaranteed not to be read, since tail hasn't been updated. Clearing the field
```

Comment:
Maybe enhance the comment to say that this is due to the logic in the `TaskRunner`. Because of the comment mentioning volatile 3 lines up , I thought you are refering to the JMM here (which would be wrong).